File: app/src/main/java/com/aiepoissac/busapp/core/data/busarrival/BusService.java
package com.aiepoissac.busapp.core.data.busarrival;

import org.json.simple.JSONObject;

import java.text.ParseException;
import java.time.format.DateTimeParseException;

final class BusService {

    private final String serviceNo;

    private final String operator;

    private final Bus[] buses;

    BusService(final JSONObject data) {
        this.serviceNo = (String) data.get("ServiceNo");
        this.operator = (String) data.get("Operator");
        this.buses = new Bus[3];
        try {
            this.buses[0] = new Bus((JSONObject) data.get("NextBus"));
        } catch (DateTimeParseException | NullPointerException e) {
            //there is guaranteed to be at least 1 next bus, for this constructor to be executed
            throw new IllegalArgumentException("Invalid Data");
        }
        try {
            this.buses[1] = new Bus((JSONObject) data.get("NextBus2"));
        } catch (DateTimeParseException | NullPointerException e) {
            this.buses[1] = null;
        }
        try {
            this.buses[2] = new Bus((JSONObject) data.get("NextBus3"));
        } catch (DateTimeParseException | NullPointerException e) {
            this.buses[2] = null;
        }
    }

    @Override
    public String toString() {
        StringBuilder s = new StringBuilder("Service: ");
        s.append(this.serviceNo);
        s.append("\t Operator: ");
        s.append(this.operator);
        s.append("\t To ");
        s.append(this.buses[0].getDestinationCode());
        s.append("\n");
        for (Bus bus : buses) {
            if (bus != null) {
                s.append(bus);
                s.append("\n");
            }
        }
        return s.toString();
    }

}

File: app/src/main/java/com/aiepoissac/busapp/data/busarrival/BusService.kt
package com.aiepoissac.busapp.data.busarrival

import com.aiepoissac.busapp.data.businfo.BusRepository
import com.aiepoissac.busapp.data.businfo.BusStopInfo
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * This class represents a bus service operating at a bus stop.
 *
 * @param serviceNo Bus service number
 * @param operator Public Transport Operator Codes:
 * • SBST (for SBS Transit)
 * • SMRT (for SMRT Corporation)
 * • TTS (for Tower Transit Singapore)
 * • GAS (for Go Ahead Singapore)
 * @param nextBus The next incoming bus. This bus is guaranteed to be valid (operating).
 * @param nextBus2 The next next incoming bus.
 * @param nextBus3 The next next next incoming bus.
 */
@Serializable
data class BusService(
    @SerialName("ServiceNo") val serviceNo: String,
    @SerialName("Operator") val operator: String,
    @SerialName("NextBus") val nextBus: Bus,
    @SerialName("NextBus2") val nextBus2: Bus,
    @SerialName("NextBus3") val nextBus3: Bus
) {

    /**
     * Returns a data structure that combines the bus arrival data of a bus service with
     * the origin and destination BusStopInfo
     *
     * @param busRepository the repository of the bus data
     */
    suspend fun attachOriginDestinationBusStopInfo(
        busRepository: BusRepository
    ): Pair<Pair<BusStopInfo?, BusStopInfo?>, BusService> {
        return Pair(
            first = Pair(
                first = nextBus.getOriginBusStopInfo(busRepository),
                second = nextBus.getDestinationBusStopInfo(busRepository)
            ),
            second = this
        )
    }
}
